File: javascript/sorting/1-BubbleSort.js
// Implement the Bubble Sort algorithm.

function bubbleSort(arr) {
  const n = arr.length;

  for (let i = 0; i < n - 1; i++) {
    // A flag to optimize the sorting process
    let swapped = false;

    for (let j = 0; j < n - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        // Swap elements if they are in the wrong order
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
        swapped = true;
      }
    }

    // If no two elements were swapped in the inner loop,
    // the array is already sorted
    if (!swapped) {
      break;
    }
  }

  return arr;
}

// Example usage
const unsortedArray = [64, 25, 12, 22, 11];
const sortedArray = bubbleSort(unsortedArray);
console.log(sortedArray); // Expected output: [11, 12, 22, 25, 64]

File: python/sorting/1-BubbleSort.py
# The Bubble Sort algorithm

def bubble_sort(arr):
  n = len(arr)

  # Traverse through all array elements
  for i in range(n):

    # Last i elements are already in place
    for j in range(0, n - i - 1):

      # Swap if the element found is greater than the next elemnt
      if (arr[j] > arr[j + 1]):
        arr[j], arr[j + 1] = arr[j + 1], arr[j]

# Test rhw bubble_sort function
arr = [64, 34, 25, 12, 22, 11, 90]
print("The original array:", arr)

bubble_sort(arr)

print("The sorted array:", arr)